Source/WebCore/rendering/RenderLayer.cpp

```diff
@@ -828,22 +828,33 @@ void RenderLayer::rebuildZOrderLists(std::unique_ptr<Vector<RenderLayer*>>& posZ
     }
 }
 
-void RenderLayer::removeSelfAndDescendantsFromCompositor()
+void RenderLayer::removeSelfFromCompositor()
 {
     if (parent())
         compositor().layerWillBeRemoved(*parent(), *this);
     clearBacking();
+}
 
-    for (RenderLayer* child = firstChild(); child; child = child->nextSibling())
-        child->removeSelfAndDescendantsFromCompositor();
+void RenderLayer::removeDescendantsFromCompositor()
+{
+    for (RenderLayer* child = firstChild(); child; child = child->nextSibling()) {
+        child->removeSelfFromCompositor();
+        child->removeDescendantsFromCompositor();
+    }
 }
 
 void RenderLayer::setWasOmittedFromZOrderTree()
 {
     if (m_wasOmittedFromZOrderTree)
         return;
 
-    removeSelfAndDescendantsFromCompositor();
+    ASSERT(!isNormalFlowOnly());
+    removeSelfFromCompositor();
+
+    // Omitting a stacking context removes the whole subtree, otherwise collectLayers will
+    // visit and omit/include descendants separately.
+    if (isStackingContext())
+        removeDescendantsFromCompositor();
```

Comment:
Doing a recursive walk of all descendant layers, while we're in the middle of a layer traversal in `collectLayers()` isn't great.

Reply:
I guess not, but it should be relatively rare, right?

This only would be slow when `visibility:hidden` is set on a stacking context, and there's large layer subtrees newly hidden.

We need to do something to make sure the compositor doesn't have any references to those layers, and the previous code was doing the same.